File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.java
class Solution {
    // Lists to store the adjacency list representation of the tree
    private List<Integer>[] adj;
    private int[] values;
    private int k;
    private int components;

    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        this.values = values;
        this.k = k;
        this.components = 0;
        
        // Initialize adjacency list
        adj = new List[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        // Build undirected tree
        for (int[] edge : edges) {
            adj[edge[0]].add(edge[1]);
            adj[edge[1]].add(edge[0]);
        }
        
        // Start DFS from root (0)
        dfs(0, -1);
        
        return components;
    }
    
    // Returns sum of values in subtree
    private long dfs(int node, int parent) {
        long sum = values[node];
        
        // Process all children except parent
        for (int child : adj[node]) {
            if (child != parent) {
                sum += dfs(child, node);
            }
        }
        
        // If sum is divisible by k, we can make this a separate component
        if (sum % k == 0) {
            components++;
            return 0;
        }
        
        return sum;
    }
}

File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.c
void addEdge(int** adj, int* sizes, int u, int v) {
    adj[u][sizes[u]++] = v;
    adj[v][sizes[v]++] = u;
}

long long dfs(int node, int parent, int* values, int** adj, int* sizes, int k, int* components) {
    long long sum = values[node];
    
    // Process all children
    for (int i = 0; i < sizes[node]; i++) {
        int child = adj[node][i];
        if (child != parent) {
            sum += dfs(child, node, values, adj, sizes, k, components);
        }
    }
    
    // If divisible by k, increment components
    if (sum % k == 0) {
        (*components)++;
        return 0;
    }
    
    return sum;
}

int maxKDivisibleComponents(int n, int** edges, int edgesSize, int* edgesColSize, 
                          int* values, int valuesSize, int k) {
    // Allocate memory for adjacency list
    int** adj = (int**)malloc(n * sizeof(int*));
    int* sizes = (int*)calloc(n, sizeof(int));
    
    // First pass to count neighbors
    for (int i = 0; i < edgesSize; i++) {
        sizes[edges[i][0]]++;
        sizes[edges[i][1]]++;
    }
    
    // Allocate memory for each vertex's neighbors
    for (int i = 0; i < n; i++) {
        adj[i] = (int*)malloc(sizes[i] * sizeof(int));
        sizes[i] = 0;  // Reset sizes for second pass
    }
    
    // Second pass to build adjacency list
    for (int i = 0; i < edgesSize; i++) {
        addEdge(adj, sizes, edges[i][0], edges[i][1]);
    }
    
    int components = 0;
    dfs(0, -1, values, adj, sizes, k, &components);
    
    // Free allocated memory
    for (int i = 0; i < n; i++) {
        free(adj[i]);
    }
    free(adj);
    free(sizes);
    
    return components;
}